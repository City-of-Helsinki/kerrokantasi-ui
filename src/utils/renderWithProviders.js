/* eslint react/prop-types: 0 */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { LoginProvider } from 'hds-react';
import { IntlProvider } from 'react-intl';

import createStore from "../createStore";
import { userOidcConfig, apiTokenClientConfig } from './oidcConfig';

const loginProviderConfig = {
  userManagerSettings: userOidcConfig,
  apiTokensClientSettings: apiTokenClientConfig,
  sessionPollerSettings: { pollIntervalInMs: 10000 },
}

const renderWithProviders = (
  ui,
  { preloadedState = {}, locale = 'fi', history, store = createStore(preloadedState), ...renderOptions } = {},
) => {
  const Wrapper = ({ children }) => (
    <LoginProvider {...loginProviderConfig}>
      <IntlProvider locale={locale}>
        <Provider store={store}>{children}</Provider>
      </IntlProvider>
    </LoginProvider>
  );
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};


export default renderWithProviders;
