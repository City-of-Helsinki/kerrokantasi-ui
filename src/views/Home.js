import React from 'react';
import PropTypes from 'prop-types';
import Helmet from 'react-helmet';
import {injectIntl, intlShape, FormattedMessage} from 'react-intl';
import {connect} from 'react-redux';
import {fetchHearingList} from '../actions';
import {getTopHearing, getOpenHearings} from '../selectors/hearing';
import {getUser} from '../selectors/user';
import {Col, Grid, Row} from 'react-bootstrap';
import FullWidthHearing from '../components/FullWidthHearing';
import HearingCardList from '../components/HearingCardList';
import orderBy from 'lodash/orderBy';
import OverviewMap from '../components/OverviewMap';
import Link from '../components/LinkWithLang';
import trackLink from '../utils/trackLink';
import CreateHearingButton from '../components/Hearings/CreateHearingButton';
import { isAdmin } from '../utils/user';
import config from '../config';
import {getFeedbackUrl} from '../utils/languageUtils';

export class Home extends React.Component {
  constructor(props) {
    super(props);

    this.state = {isMobile: typeof window !== 'undefined' && window.innerWidth < 768};
    this.handleResize = this.handleResize.bind(this);
  }
  /**
   * Return a promise that will, as it fulfills, have added requisite
   * data for the home view into the dispatch's associated store.
   *
   * @param dispatch Redux Dispatch function
   * @return {Promise} Data fetching promise
   */
  static fetchData(dispatch) {
    return Promise.all([
      dispatch(fetchHearingList('topHearing', '/v1/hearing', {ordering: '-n_comments', open: true, limit: 1})),
      dispatch(fetchHearingList('openHearings', '/v1/hearing', {open: true, include: 'geojson'})),
    ]);
  }

  componentDidMount() {
    Home.fetchData(this.props.dispatch);
    if (typeof window !== 'undefined') window.addEventListener('resize', this.handleResize);
    trackLink();
  }

  componentWillUnmount() {
    if (typeof window !== 'undefined') window.removeEventListener('resize', this.handleResize);
  }

  handleResize() {
    this.setState({isMobile: typeof window !== 'undefined' && window.innerWidth < 768});
  }

  render() {
    const {formatMessage} = this.props.intl;
    const {topHearing, openHearings, language, user, intl} = this.props;
    const {isMobile} = this.state;
    const hearingMap =
      openHearings && openHearings.data ? (
        <div className="map">
          <h2>
            <FormattedMessage id="open-hearings-on-map" />
          </h2>
          <OverviewMap
            enablePopups
            hearings={openHearings.data}
            style={{width: '100%', height: isMobile ? '70%' : 600}}
            mapElementLimit={10}
          />
        </div>
      ) : null;
    const heroStyle = {
      backgroundImage: "url('" + config.heroImageURL + "')",
    };

    return (
      <div>
        <section className="page-section page-section--welcome" style={heroStyle}>
          <Grid>
            <Row>
              <Col xs={10} md={8} className="welcome-content">
                <Helmet
                  title={formatMessage({id: 'welcome'})}
                  meta={[{name: "description", content: formatMessage({id: 'descriptionTag'})}]}
                />
                <h1>
                  <FormattedMessage id="welcome" />
                </h1>
                <p className="lead">
                  <FormattedMessage id="welcomeMessage" />
                </p>
              </Col>
            </Row>
          </Grid>
          <div className="welcome-koro__bottom" />
        </section>
        <section className="page-section page-section--hearing-card">
          <div className="container">
            <h2 className="page-title">
              <FormattedMessage id="openHearings" />
            </h2>
            {topHearing && <FullWidthHearing hearing={topHearing} language={language} intl={intl}/>}
            {openHearings && openHearings.data && topHearing &&
              !openHearings.isFetching && (
                <div className="list">
                  <HearingCardList
                    hearings={
                      orderBy(openHearings.data
                        .filter(hearing => hearing.id !== topHearing.id), ['close_at'], ['desc'])
                    }
                    language={language}
                    intl={intl}
                  />
                  <p className="text-center">
                    <Link to={{path: "/hearings/list"}} className="btn btn-default">
                      <FormattedMessage id="allHearings" />
                    </Link>
                  </p>
                </div>
              )}
          </div>
          <div className="hearings-koro__bottom" />
        </section>
        <section className="page-section page-section--feedback">
          <div className="container">
            <h2 className="feedback-prompt">
              <a href={getFeedbackUrl(language)} target="_blank" rel="noopener noreferrer" className="feedback-box">
                <FormattedMessage id="feedbackPrompt" />{' '}
              </a>
            </h2>
          </div>
        </section>
        <section className="page-section page-section--map">
          <div className="container">
            {hearingMap}
          </div>
        </section>
        {isAdmin(user) && <CreateHearingButton to={{path: '/hearing/new'}} />}
      </div>
    );
  }
}

Home.propTypes = {
  intl: intlShape.isRequired,
  dispatch: PropTypes.func,
  openHearings: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  topHearing: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  language: PropTypes.string, // make sure changing language refreshes
  user: PropTypes.object,
};

const mapStateToProps = state => ({
  topHearing: getTopHearing(state),
  language: state.language,
  openHearings: getOpenHearings(state),
  user: getUser(state),
});

const WrappedHome = connect(mapStateToProps)(injectIntl(Home));
// We need to re-hoist the static fetchData to the wrapped component due to react-intl:
WrappedHome.fetchData = Home.fetchData;
export default WrappedHome;
