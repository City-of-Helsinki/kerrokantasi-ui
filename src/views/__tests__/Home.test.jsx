import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { screen } from '@testing-library/react';

import { Home } from '../Home';
import { getIntlAsProp, mockStore } from '../../../test-utils';
import renderWithProviders from '../../utils/renderWithProviders';

const renderComponent = (propsOverrides) => {
  const {
    hearingLists: { allHearings },
    ...rest
  } = mockStore;

  const props = {
    topHearing: allHearings.data[0],
    openHearings: allHearings,
    ...rest,
    ...propsOverrides,
  };

  return renderWithProviders(
    <MemoryRouter>
      <Home intl={getIntlAsProp()} {...props} />
    </MemoryRouter>,
  );
};

describe('<Home />', () => {
  it('should render correctly', () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it('should find top hearing', () => {
    renderComponent();

    expect(screen.getByText('Mock Hearing')).toBeInTheDocument();
  });

  it('should find HearingCardList', async () => {
    renderComponent();

    expect(await screen.findByTestId('hearing-card-list')).toBeInTheDocument();
  });
});
