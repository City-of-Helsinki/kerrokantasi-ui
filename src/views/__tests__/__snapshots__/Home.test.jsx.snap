// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render correctly 1`] = `
<div>
  <div>
    <section
      class="page-section page-section--welcome"
      style="background-image: url(http://materialbank.myhelsinki.fi/detail/1192/download/7);"
    >
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="welcome-content col-md-8 col-xs-10"
          >
            <h1>
              welcome
            </h1>
            <p
              class="lead"
            >
              welcomeMessage
            </p>
          </div>
        </div>
      </div>
      <div
        class="welcome-koro__bottom"
      />
    </section>
    <section
      class="page-section page-section--hearing-card"
    >
      <div
        class="container"
      >
        <h2
          class="page-title"
        >
          openHearings
        </h2>
        <div
          class="fullwidth-hearing "
        >
          <div
            alt="A Wonderfull caption"
            class="mouse-only-click-element fullwidth-hearing-image"
            style="background-image: url(https://api.hel.fi/kerrokantasi-test/media/images/2017/10/LzqVMrMI.jpeg);"
          />
          <div
            class="fullwidth-hearing-header"
          >
            <div
              class="fullwidth-hearing-title-wrap"
            >
              <h3
                class="h2 fullwidth-hearing-title"
              >
                <a
                  href="/mockHearing?headless=false"
                >
                  Mock Hearing
                </a>
              </h3>
            </div>
            <div
              class="fullwidth-hearing-comments"
            >
              <span
                aria-hidden="true"
                class="fa fa-comment-o"
              />
              Â 
              1
              <span
                class="sr-only"
              >
                hearingCardFullWidthComment
              </span>
            </div>
            <div
              class="fullwidth-hearing-times"
            >
              <span>
                timeOpenPast
                 
                8 vuotta sitten
                .
              </span>
               
              <div>
                timeCloseFutureWithValues
                .
              </div>
            </div>
            <div
              class="fullwidth-hearing-labels clearfix"
            >
              <div
                aria-label="labelListTagsTitle"
                class="hearing-list-item-labellist"
              />
            </div>
            <div
              class="hearing-card-notice"
            >
              <span
                aria-hidden="true"
                class="fa fa-exclamation-circle"
              />
              hearingTranslationNotAvailable
              <div
                class="language-available-message"
                lang="fi"
              >
                Kuuleminen saatavilla suomeksi
              </div>
            </div>
          </div>
        </div>
        <div
          class="list"
        >
          <div
            class="hearing-card-list row"
            data-testid="hearing-card-list"
          />
          <p
            class="text-center"
          >
            <a
              href="/hearings/list?headless=false"
            >
              <button
                class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO kerrokantasi-btn"
                type="button"
              >
                <span
                  class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
                >
                  allHearings
                </span>
              </button>
            </a>
          </p>
        </div>
      </div>
      <div
        class="hearings-koro__bottom"
      />
    </section>
    <section
      class="page-section page-section--map"
    >
      <div
        class="container"
      >
        <div
          class="map"
        >
          <h2>
            open-hearings-on-map
          </h2>
          <div
            data-testid="MapContainer"
          >
            <div
              data-testid="TileLayer"
            />
            <div
              data-testid="FeatureGroup"
            />
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
